### DISCUSSION TITLE
Music playback progress by percentage

### DISCUSSION THREAD

**Discussion by BeijingUncle** [pschatzmann/arduino-audio-tools#630] [Q&A] [answered]
Hi, Phill:

I'm trying to make a mp3 player and for some reason I can not use audioPlayer class.

I then start with `streams-sd_mp3-metadata` example.

I have to determine playback is finished and playback progress. In StreamCopy class I found a callback `setMimeCallback` and use it with `available()`, then I can do both well.

My question is: Is this a right(normal) way, and is there a better way to do progress thing?

**Comment by pschatzmann** [comment, marked as answer]
Currently there is no such functionality, but I suggest that you contribute to this project by extending it. 

I think the cleanest way would be to provide a new class e.g. ProgressStream() where you implement the following methods:
- setInput(Stream&)
- setInput(File&): calls setInput() and setSize()
- setOutput(Print&)
- setSize(size_t size) updates the total size and resets the bytes consumed
- bytesConsumed() provides the bytes consumed which are updated by the write() and readBytes() methods
- percentage(): calculated from size and bytes consumed
- write() writes the provided bytes to the final output
- readBytes provides the bytes from the final input

Maybe you have a look at the existing MeasuringStream class which calculating the bytes per second...

Of cause you can easily do this calculation in the loop by just calculating the percentage using file.position() and file.size() and there is no need for a callback.

**Comment by BeijingUncle** [comment]
I do think i'm not good enough to do this, maybe I can do some testing stuff, sorry about that.

**Comment by pschatzmann** [comment]
Have a look at https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_progress_stream.html

**Comment by BeijingUncle** [comment]
Wow! What a speed! Thank you!
